lib/iris/analysis/__init__.py

```diff
@@ -506,22 +528,270 @@ def aggregate(self, data, axis, **kwargs):
+
+        """
+        if isinstance(data_result, biggus.Array):
+            collapsed_cube.lazy_data(data_result)
+        else:
+            collapsed_cube.data = data_result
+
+        return collapsed_cube
+
+    def aggregate_shape(self, **kwargs):
+        """
+        The shape of the additive dimension created by the aggregator.
+
+        Kwargs:
+
+        * This function is intended to be used in conjunction with aggregate()
+          and should be passed the same keywords.
+
+        Returns:
+            A tuple of the additive dimension shape.
+
+        """
+        return ()
+
+    @abstractmethod
+    def name(self):
+        """
+        Returns the name of the aggregator.
+
+        """
+
+
+class AdditiveAggregator(_Aggregator):
+    """
+    The :class:`AdditiveAggregator` class provides common additive aggregation
+    functionality.
+
+    An additive aggregator *may* introduce a new dimension to the data for the
+    statistic being calculated, but only if more than one point is requested.
+    For example, calculating the 50th and 90th percentile will result in a new
+    data dimension with an extent of 2, for each of the quantiles calculated.
+
+    """
+
+    def __init__(self, name, call_func, required, units_func=None,
+                 lazy_func=None, **kwargs):
+        """
+        Create an additive aggregator for the given :data:`call_func`.
+
+        Args:
+
+        * name (string):
+            The name of the aggregator.
+
+        * call_func (callable):
+            | *Call signature*: (data, axis=None, \**kwargs)
+
+            Data aggregation function.
+            Returns an aggregation result, collapsing the 'axis' dimension of
+            the 'data' argument.
+
+        * required (string):
+            The name of the mandatory keyword argument associated with
+            calling this aggregator.
+
+        Kwargs:
+
+        * units_func (callable):
+            | *Call signature*: (units)
+
+            If provided, called to convert a cube's units.
+            Returns an :class:`iris.units.Unit`, or a
+            value that can be made into one.
+
+        * lazy_func (callable or None):
+            An alternative to :data:`call_func` implementing a lazy
+            aggregation. Note that, it need not support all features of the
+            main operation, but should raise an error in unhandled cases.
+
+        Additional kwargs::
+            Passed through to :data:`call_func` and :data:`lazy_func`.
+
+        Aggregators are used by cube aggregation methods such as
+        :meth:`~iris.cube.Cube.collapsed` and
+        :meth:`~iris.cube.Cube.aggregated_by`.  For example::
+
+            cube.collapsed('longitude', iris.analysis.PERCENTILE, percent=50)
+
+        """
+        self._name = name
+        if not isinstance(required, basestring) or not required:
+            msg = '{} aggregator requires a mandatory keyword argument.'
```

Comment:
Is it likely that user code will generate this error? Or is it expected that this class will only be initialised in a controlled manner such that the required argument will always be passed.

I think if I received this error that I would be very confused indeed: what mandatory kwarg? What's it called? What does it do? Where and when should I pass it, and what value should it take?

I notice that earlier a very similar error message contains the name of the missing kwarg; perhaps including that would help here too :smile: